### PR TITLE
[SPARK-48220][PYTHON] Allow passing PyArrow Table to createDataFrame()

### PR DESCRIPTION
**Pull request by ianmcook**
### What changes were proposed in this pull request?
- Add support for passing a PyArrow Table to `createDataFrame()`.
- Document this on the **Apache Arrow in PySpark** user guide page.
- Fix an issue with timestamp and struct columns in `toArrow()`.

### Why are the changes needed?
This seems like a logical next step after the addition of a `toArrow()` DataFrame method in #45481.

### Does this PR introduce _any_ user-facing change?
Users will have the ability to pass PyArrow Tables to `createDataFrame()`. There are no changes to the parameters of `createDataFrame()`. The only difference is that `data` can now be a PyArrow Table.

### How was this patch tested?
Many tests were added, for Spark Classic and Spark Connect. I ran the tests locally with older versions of PyArrow installed (going back to 10.0).

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: python/pyspark/sql/pandas/conversion.py [hunk 1]
```diff
@@ -360,42 +361,52 @@ def createDataFrame(  # type: ignore[misc]
 
         assert isinstance(self, SparkSession)
```
Comment (ianmcook) [inline python/pyspark/sql/pandas/conversion.py:363]: The git diff makes it difficult to see what I did here. I just moved a few lines around at the top of this function and added conditional blocks below:

```py
if isinstance(data, pa.Table):
  # some new code here
else:
  # the existing code here
```

File: python/pyspark/sql/tests/typing/test_session.yml [hunk 2]
```diff
@@ -51,25 +51,6 @@
     spark.createDataFrame(["foo", "bar"], "string")
 
 
-- case: createDataFrameScalarsInvalid
```
Comment (ianmcook) [inline python/pyspark/sql/tests/typing/test_session.yml:54]: Because of PyArrow's lack of type annotations, the changes in this PR caused this typing test to fail.

However, the failure modes checked in this test are adequately caught without relying on PySpark's type annotations.

For the first case, PySpark raises:
```
pyspark.errors.exceptions.base.PySparkTypeError: [CANNOT_INFER_SCHEMA_FOR_TYPE] Can not infer schema for type: `str`.
```

For the second case, PySpark raises:
```
pyspark.errors.exceptions.base.PySparkTypeError: [FIELD_DATA_TYPE_UNACCEPTABLE] StructType([StructField('name', StringType(), True), StructField('age', IntegerType(), True)]) can not accept object 1 in type <class 'int'>.
```

Given that the errors are caught by other means, it seems reasonable to remove this test.

File: python/pyspark/sql/session.py [hunk 3]
```diff
@@ -1251,7 +1252,7 @@ def _getActiveSessionOrCreate(**static_conf: Any) -> "SparkSession":
             spark = builder.getOrCreate()
         return spark
 
-    @overload
+    @overload  # type: ignore[override]
```
Comment (ianmcook) [inline python/pyspark/sql/session.py:1255]: PyArrow's lack of type annotations makes this ignore necessary. I tried a variety of different ways to avoid adding it, with no success.

File: python/pyspark/sql/connect/session.py [hunk 4]
```diff
@@ -474,35 +475,40 @@ def createDataFrame(
 
         _table: Optional[pa.Table] = None
 
-        if isinstance(data, pd.DataFrame):
+        if isinstance(data, pd.DataFrame) or isinstance(data, pa.Table):
             # Logic was borrowed from `_create_from_pandas_with_arrow` in
             # `pyspark.sql.pandas.conversion.py`. Should ideally deduplicate the logics.
 
             # If no schema supplied by user then get the names of columns only
             if schema is None:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/session.py:483]: quick comment, I wonder if we can flatten this if-else via:

```python
if schema is None and isinstance(data, pd.DataFrame):
    ...
elif schema is None and  isinstance(data, pa.Table)
    ...
```

Then less conflicts to other branches if we land some fixes in the future, easier to backport/revert, etc.

File: python/pyspark/sql/session.py (hunk 3 above)
Comment (amol-) [reply python/pyspark/sql/session.py:1255]: Are you sure this isn't due to the fact that the provided overload doesn't have the `schema` argument? Overriding a method while adding or removing arguments will generally trigger an error in mypy (which override does mute)

Comment (ianmcook) [reply python/pyspark/sql/session.py:1255]: Yes, I didn't change anything involving the `schema` argument. I think the problem is that PyArrow classes appear in mypy type check results as `Any`, as described at https://github.com/apache/arrow/issues/24376#issuecomment-1377869384. The type checker doesn't like this.

File: python/pyspark/sql/connect/session.py (hunk 4 above)
Comment (ianmcook) [reply python/pyspark/sql/connect/session.py]: I flattened the conditionals. Please take a look: https://github.com/apache/spark/pull/46529/commits/795d01a1fd96fcbe80432d74be50382d350edf57?diff=split&w=0

**Comment by alippai** [comment]
This makes the usage so much easier, thanks!

What will happen with the nanosecond timestamps? Truncated to milliseconds?

**Comment by ianmcook** [comment]
> What will happen with the nanosecond timestamps? Truncated to milliseconds?

Yes, truncated to milliseconds for now.

**Comment by alippai** [comment]
I wish we had ns support in spark 4.0 as Java, parquet, arrow etc uses it natively now, but that’s certainly a different discussion and MRs :)

Thanks for the new API!

File: python/pyspark/sql/tests/connect/test_parity_arrow.py [hunk 5]
```diff
@@ -31,8 +31,17 @@ def test_createDataFrame_fallback_disabled(self):
...
 
-    def test_createDataFrame_with_map_type(self):
-        self.check_createDataFrame_with_map_type(True)
+    def test_createDataFrame_pandas_with_map_type(self):
+        self.check_createDataFrame_pandas_with_map_type(True)
+
+    def test_createDataFrame_pandas_with_struct_type(self):
+        self.check_createDataFrame_pandas_with_struct_type(True)
+
+    def test_createDataFrame_arrow_with_struct_type(self):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_parity_arrow.py:40]: We can remove this if that takes no argument. It will be inherited, and run the tests

**Comment by HyukjinKwon** [comment]
cc @zhengruifeng @ueshin @xinrong-meng FYI

File: python/pyspark/sql/tests/connect/test_parity_arrow.py (hunk 5 above)
Comment (ianmcook) [reply python/pyspark/sql/tests/connect/test_parity_arrow.py]: Ah, right. Done in 292d3c8. Thanks.

File: python/pyspark/sql/tests/typing/test_session.yml (hunk 2 above)
Comment (ianmcook) [reply python/pyspark/sql/tests/typing/test_session.yml:54]: I added tests in 40072d0 to ensure we still cover these error conditions somewhere in the tests.

File: python/pyspark/sql/connect/session.py [hunk 6]
```diff
@@ -555,11 +575,21 @@ def createDataFrame(
                 ]
             )
 
-            if isinstance(schema, StructType):
-                assert arrow_schema is not None
-                _table = _table.rename_columns(
-                    cast(StructType, _deduplicate_field_names(schema)).names
-                ).cast(arrow_schema)
+        elif isinstance(data, pa.Table):
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/session.py:578]: Are above changes in this file related to this PR?

File: python/pyspark/sql/context.py [hunk 7]
```diff
@@ -46,6 +46,7 @@
 
 if TYPE_CHECKING:
     from py4j.java_gateway import JavaObject
+    import pyarrow as pa
```
Comment (zhengruifeng) [inline python/pyspark/sql/context.py:49]: I think we can not assume `pyarrow` is installed by default in Spark Classic

https://spark.apache.org/docs/latest/api/python/getting_started/install.html#dependencies

@HyukjinKwon

File: python/pyspark/sql/context.py [hunk 8]
```diff
@@ -343,28 +344,29 @@ def createDataFrame(
 
     @overload
     def createDataFrame(
-        self, data: "PandasDataFrameLike", samplingRatio: Optional[float] = ...
+        self, data: Union["PandasDataFrameLike", "pa.Table"], samplingRatio: Optional[float] = ...
```
Comment (zhengruifeng) [inline python/pyspark/sql/context.py:347]: can we include `pa.Table` in `PandasDataFrameLike`?

File: python/pyspark/sql/pandas/types.py [hunk 9]
```diff
@@ -162,6 +243,8 @@ def from_arrow_type(at: "pa.DataType", prefer_timestamp_ntz: bool = False) -> Da
         spark_type = StringType()
     elif types.is_binary(at):
         spark_type = BinaryType()
+    elif types.is_fixed_size_binary(at):
+        spark_type = BinaryType()
```
Comment (zhengruifeng) [inline python/pyspark/sql/pandas/types.py:247]: is it related?

File: python/pyspark/sql/pandas/types.py [hunk 10]
```diff
@@ -174,6 +257,18 @@ def from_arrow_type(at: "pa.DataType", prefer_timestamp_ntz: bool = False) -> Da
         spark_type = DayTimeIntervalType()
     elif types.is_list(at):
         spark_type = ArrayType(from_arrow_type(at.value_type, prefer_timestamp_ntz))
+    elif types.is_fixed_size_list(at):
+        import pyarrow as pa
+
+        if LooseVersion(pa.__version__) < LooseVersion("14.0.0"):
```
Comment (zhengruifeng) [inline python/pyspark/sql/pandas/types.py:263]: we'd better support version >= 4.0.0

https://spark.apache.org/docs/latest/api/python/getting_started/install.html#dependencies

**Review by zhengruifeng — COMMENTED** [review]
This PR is too big to follow and has many unrelated changes, can we break it to multiple PRs ?

File: python/pyspark/sql/context.py (hunk 7 above)
Comment (ianmcook) [reply python/pyspark/sql/context.py:49]: I think it is OK inside when `TYPE_CHECKING` is True. There is another file in the repo that has had `import pyarrow as pa` inside an `if TYPE_CHECKING:` conditional since 2021: https://github.com/apache/spark/pull/34101/files#diff-a4f1631a18d1b4921b8727e1d78059a1014c433239eb107962c473c6466214e5R29

**Comment by ianmcook** [comment]
> This PR is too big to follow and has many unrelated changes, can we break it to multiple PRs ?

Yes I am happy to break out the non-required changes into separate PRs.

File: python/pyspark/sql/pandas/types.py (hunk 9 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/types.py:247]: It's not required. I will move this to a separate PR.

File: python/pyspark/sql/context.py (hunk 8 above)
Comment (ianmcook) [reply python/pyspark/sql/context.py:347]: I don't think so. `PandasDataFrameLike` is just an alias for `pd.DataFrame`.

In `pyspark.sql.pandas._typing`, `DataFrameLike` is defined like this:
```py
from pandas.core.frame import DataFrame as PandasDataFrame
DataFrameLike = PandasDataFrame
```

Then in other parts of PySpark, `PandasDataFrameLike` is defined like this:
```py
from pyspark.sql.pandas._typing import DataFrameLike as PandasDataFrameLike
```

If we define `PandasDataFrameLike` as a Union that includes `pa.Table`, that will cause other problems. For example, then we can't use it as the return type of `toPandas()`.

File: python/pyspark/sql/connect/session.py (hunk 6 above)
Comment (ianmcook) [reply python/pyspark/sql/connect/session.py:578]: I made these changes because Hyukjin asked me to flatten the if-else logic in this function. See https://github.com/apache/spark/pull/46529#discussion_r1597744517. But I think this makes the PR difficult to review. I will revert this and we can consider doing it in a follow-on PR.

File: python/pyspark/sql/pandas/types.py (hunk 9 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/types.py]: Moved to #46688

File: python/pyspark/sql/connect/session.py (hunk 6 above)
Comment (ianmcook) [reply python/pyspark/sql/connect/session.py:571]: Ok, I was able to refactor this to make it much simpler to review.

**Comment by ianmcook** [comment]
@zhengruifeng I broke out the non-required changes into separate PRs and I simplified the implementations. I think it should be easier to follow now. More than half of the added code is tests.

File: python/pyspark/sql/connect/dataframe.py [hunk 11]
```diff
@@ -1770,8 +1770,9 @@ def _to_table(self) -> Tuple["pa.Table", Optional[StructType]]:
         return (table, schema)
 
     def toArrow(self) -> "pa.Table":
+        schema = to_arrow_schema(self.schema, error_on_duplicated_field_names_in_struct=True)
         table, _ = self._to_table()
-        return table
+        return table.cast(schema)
```
Comment (ianmcook) [inline python/pyspark/sql/connect/dataframe.py:1775]: This fixes issues with timestamp and struct columns which were introduced in #45481.

File: python/pyspark/sql/pandas/conversion.py [hunk 12]
```diff
@@ -236,15 +236,15 @@ def toArrow(self) -> "pa.Table":
...
+        schema = to_arrow_schema(self.schema, error_on_duplicated_field_names_in_struct=True)
 
         import pyarrow as pa
 
         self_destruct = jconf.arrowPySparkSelfDestructEnabled()
         batches = self._collect_as_arrow(
             split_batches=self_destruct, empty_list_if_zero_records=False
         )
-        table = pa.Table.from_batches(batches)
+        table = pa.Table.from_batches(batches).cast(schema)
```
Comment (ianmcook) [inline python/pyspark/sql/pandas/conversion.py:247]: This fixes issues with timestamp and struct columns which were introduced in #45481 for Classic.

File: python/pyspark/sql/tests/test_arrow.py [hunk 13]
```diff
@@ -315,6 +346,17 @@ def check_create_data_frame_to_pandas_timestamp_ntz(self, arrow_enabled):
                 pdf = df.toPandas()
             assert_frame_equal(origin, pdf)
```
Comment (ianmcook) [inline python/pyspark/sql/tests/test_arrow.py:348]: Each Arrow test here is based on a pandas test above it.

- `test_create_data_frame_to_arrow_timestamp_ntz` is based on `test_create_data_frame_to_pandas_timestamp_ntz`
- `test_createDataFrame_arrow_respect_session_timezone` is based on `test_createDataFrame_pandas_respect_session_timezone`

... and so on. I renamed the pandas ones with `pandas` in the name for clarity.

File: python/pyspark/sql/pandas/types.py [hunk 14]
```diff
@@ -232,6 +312,124 @@ def _get_local_timezone() -> str:
...
+    return pa.Table.from_arrays(
+        [
+            _check_arrow_array_timestamps_localize(a, f.dataType, truncate, timezone)
+            for a, f in zip(table.columns, schema.fields)
+        ],
+        schema=table.schema,
+    )
+
+
+def _check_arrow_array_timestamps_localize(
```
Comment (ianmcook) [inline python/pyspark/sql/pandas/types.py:350]: I am sorry if it is hard to follow what this function does. It looks for Arrow Timestamp columns that will be converted to Spark TIMESTAMP_LTZ columns. It localizes them (`pc.assume_timezone`) and optionally truncates them to microseconds (`pc.floor_temporal`). If it finds a chunked array, list array, map array, struct array, or dictionary array, then it recurses into the child arrays.

cc @jorisvandenbossche in case you want to take a look

**Comment by ianmcook** [comment]
Here is a PDF of the Apache Arrow in PySpark user guide page rendered from this PR:
[Apache Arrow in PySpark — PySpark master documentation.pdf](https://github.com/apache/spark/files/15395322/Apache.Arrow.in.PySpark.PySpark.master.documentation.pdf)

File: python/pyspark/sql/tests/test_arrow.py [hunk 15]
```diff
@@ -199,13 +237,6 @@ def create_arrow_table(self):
             pa.field("6_decimal_t", pa.decimal128(38, 18)),
         )
         t = t.cast(new_schema)
-        # convert timestamp to local timezone
-        timezone = self.spark.conf.get("spark.sql.session.timeZone")
-        t = t.set_column(
-            t.schema.get_field_index("8_timestamp_t"),
-            "8_timestamp_t",
-            pc.assume_timezone(t["8_timestamp_t"], timezone),
-        )
```
Comment (ianmcook) [inline python/pyspark/sql/tests/test_arrow.py:208]: I added this in #45481. The timestamp fixes here make this test work without localizing the timezone.

File: python/pyspark/sql/connect/session.py [hunk 16]
```diff
@@ -413,11 +416,14 @@ def _inferSchemaFromList(
...
         self,
-        data: Union["pd.DataFrame", "np.ndarray", Iterable[Any]],
+        data: Union["pd.DataFrame", "np.ndarray", "pa.Table", Iterable[Any]],
         schema: Optional[Union[AtomicType, StructType, str, List[str], Tuple[str, ...]]] = None,
         samplingRatio: Optional[float] = None,
         verifySchema: Optional[bool] = None,
     ) -> "ParentDataFrame":
+        import pandas as pd
+        import pyarrow as pa
+
```
Comment (jorisvandenbossche) [inline python/pyspark/sql/connect/session.py:426]:
```suggestion
```

Those packages are already imported top-level in this file as well?

File: python/pyspark/sql/pandas/types.py (hunk 14 above)
Comment (jorisvandenbossche) [reply python/pyspark/sql/pandas/types.py:315]: Generally looks good to me. But from seeing that code, it reminds me that we should really have some kernels to apply a scalar function on all values in a nested array like a list array. Although that would only clean up the list case in the function below (not struct and dict), not sure if we want something more generic like a visitor that visits each child array and applies some function on it.

File: python/pyspark/sql/pandas/types.py [hunk 17]
```diff
@@ -232,6 +312,124 @@ def _get_local_timezone() -> str:
...
+
+        # Only localize timestamps that will become Spark TimestampType columns.
+        # Do not localize timestamps that will become Spark TimestampNTZType columns.
+        return pc.assume_timezone(a, timezone)
+    if types.is_list(a.type):
+        at: ArrayType = cast(ArrayType, dt)
+        return pa.ListArray.from_arrays(
+            a.offsets,
+            _check_arrow_array_timestamps_localize(a.values, at.elementType, truncate, timezone),
+        )
```
Comment (jorisvandenbossche) [inline python/pyspark/sql/pandas/types.py:368]: Also this will not preserve nulls? (similarly like the issue you raised for map type, although ListArray already has a `mask` keyword. And we should also add something like https://github.com/apache/arrow/issues/23380 to simply apply an existing validity bitmap buffer)

File: python/pyspark/sql/pandas/types.py [hunk 18]
```diff
@@ -232,6 +312,124 @@ def _get_local_timezone() -> str:
...
+    if types.is_map(a.type):
+        mt: MapType = cast(MapType, dt)
+        # TODO(SPARK-48302): Do not replace nulls in MapArray with empty lists
+        return pa.MapArray.from_arrays(
+            a.offsets,
+            _check_arrow_array_timestamps_localize(a.keys, mt.keyType, truncate, timezone),
+            _check_arrow_array_timestamps_localize(a.items, mt.valueType, truncate, timezone),
+        )
+    if types.is_struct(a.type):
```
Comment (jorisvandenbossche) [inline python/pyspark/sql/pandas/types.py:377]: Similar comment here, as StructArray can have a top-level nulls / validity bitmap that will be lost here (I don't know if Spark has this ability, though)

File: examples/src/main/python/sql/arrow.py [hunk 19]
```diff
@@ -33,20 +33,23 @@
...
+    table = pa.table([pa.array(np.random.rand(100)) for i in range(3)], names=["a", "b", "c"])
 
-    # Create a Spark DataFrame
-    df = spark.range(100).drop("id").withColumns({"0": rand(), "1": rand(), "2": rand()})
+    # Create a Spark DataFrame from the PyArrow Table
+    df = spark.createDataFrame(table)
 
     # Convert the Spark DataFrame to a PyArrow Table
-    table = df.select("*").toArrow()
+    result_table = df.select("*").toArrow()
```
Comment (xinrong-meng) [inline examples/src/main/python/sql/arrow.py:47]: Out of curiosity why do we explicitly "select("*")"?

File: python/pyspark/sql/connect/session.py [hunk 20]
```diff
@@ -561,6 +568,28 @@ def createDataFrame(
...
 
+        elif isinstance(data, pa.Table):
+            prefer_timestamp_ntz = is_timestamp_ntz_preferred()
+
+            (timezone,) = self._client.get_configs("spark.sql.session.timeZone")
+
+            # If no schema supplied by user then get the names of columns only
+            if schema is None:
+                _cols = data.column_names
+            if isinstance(schema, (list, tuple)) and cast(int, _num_cols) < len(data.columns):
```
Comment (xinrong-meng) [inline python/pyspark/sql/connect/session.py:579]: nit:

```
            if isinstance(schema, (list, tuple)) and cast(int, _num_cols) < len(data.columns):
                assert isinstance(_cols, list)
                _cols.extend([f"_{i + 1}" for i in range(cast(int, _num_cols), len(data.columns))])
                _num_cols = len(_cols)
```
is uneasy to follow and duplicated, shall we extract and reuse, or add a comment to both to help understand? Feel free to do it in a separate PR

File: python/pyspark/sql/pandas/types.py (hunk 17 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/types.py]: Thanks for catching this. Fixed in 5db5e4b. Added tests in fc758e6 to fail if nulls are not preserved here.

File: python/pyspark/sql/pandas/types.py (hunk 18 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/types.py:378]: Fixed in 5db5e4b. Added tests in fc758e6 to fail if nulls are not preserved here.

File: python/pyspark/sql/pandas/types.py (hunk 17 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/types.py]: I also confirmed that it's not necessary to pass `mask` when creating the dictionary array; the nulls in the dictionary indices pass through fine and are preserved.

File: examples/src/main/python/sql/arrow.py (hunk 19 above)
Comment (ianmcook) [reply examples/src/main/python/sql/arrow.py:47]: I followed the pandas example (see below on line 69 of this same file). I was wondering this too, but I kept it just to match the pandas example. I'm happy to remove both if that would be better.

File: python/pyspark/sql/connect/session.py (hunk 20 above)
Comment (ianmcook) [reply python/pyspark/sql/connect/session.py:579]: This is borrowed directly from the pandas section above (line 516). I would be happy to open a subsequent PR to try to deduplicate the logic and make it clearer.

File: python/pyspark/sql/connect/session.py (hunk 16 above)
Comment (ianmcook) [reply python/pyspark/sql/connect/session.py]: Thanks. Removed in 73a7886.

File: examples/src/main/python/sql/arrow.py (hunk 19 above)
Comment (ianmcook) [reply examples/src/main/python/sql/arrow.py:47]: I suspect that the original purpose of the `.select("*")` was to represent some arbitrary transformations being lazily performed on the dataframe. That way users will know that this works when there are transformations.

File: python/pyspark/sql/pandas/types.py (hunk 17 above)
Comment (jorisvandenbossche) [reply python/pyspark/sql/pandas/types.py]: I don't know how much more complex you want to make this part of the code, but one disadvantage with the current `mask` keyword is that it is not zero copy (since it inverts the validity bitmap twice), so you might want to avoid having to do this as much as possible:

- you could check if the list's `value_type` is either timestamp or nested, and only in that case call `_check_arrow_array_timestamps_localize` on the value, and otherwise just return `a` (we have `pa.types.is_nested` that could help with that). That should already make any simple list of numeric type fully zero-copy
- if you do have to recreated the ListArray, you can probably do `mask=a.is_null() if a.null_count else None` to avoid allocating a full bitmap in case there are no missing values

Comment (ianmcook) [reply python/pyspark/sql/pandas/types.py]: I think it is worthwhile if it keeps zero-copy. Added in 7ed45e0. I did it for StructArrays too.

Comment (jorisvandenbossche) [reply python/pyspark/sql/pandas/types.py]: Those edits look good to me!

**Review by jorisvandenbossche — APPROVED** [review]
I haven't looked the full PR in detail, but the parts heavily involving pyarrow APIs look good to me

**Comment by ianmcook** [comment]
@zhengruifeng is there anything else I can do to help with this? Thank you

File: python/pyspark/sql/context.py (hunk 7 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/context.py:49]: yeah this is fine with TYPE_CHECKING

File: python/pyspark/sql/pandas/types.py [hunk 21]
```diff
@@ -86,30 +110,58 @@ def to_arrow_type(dt: DataType) -> "pa.DataType":
...
+        )
+        value_field = pa.field(
+            "value",
+            to_arrow_type(dt.valueType, error_on_duplicated_field_names_in_struct, timestamp_utc),
+            nullable=dt.valueContainsNull,
+        )
         arrow_type = pa.map_(key_field, value_field)
     elif type(dt) == StructType:
+        field_names = dt.names
+        if error_on_duplicated_field_names_in_struct and len(set(field_names)) != len(field_names):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/types.py:143]: Quick question. Why do we enforce this? Did it not work before?

**Review by HyukjinKwon — APPROVED** [review]
Looks pretty good. Would be good to have @BryanCutler @ueshin 's looks too.

File: python/pyspark/sql/pandas/conversion.py [hunk 22]
```diff
@@ -360,12 +376,25 @@ def createDataFrame(  # type: ignore[misc]
 
         assert isinstance(self, SparkSession)
 
+        timezone = self._jconf.sessionLocalTimeZone()
+
+        if type(data).__name__ == "Table":
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/conversion.py:381]: I wonder if there's a better way to do this.

Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/conversion.py:381]: We could check if `pyarrow` exists, and maybe just `isinstance` comparison would be nicer.

File: python/pyspark/sql/pandas/types.py [hunk 23]
```diff
@@ -124,12 +176,40 @@ def to_arrow_type(dt: DataType) -> "pa.DataType":
     return arrow_type
 
 
-def to_arrow_schema(schema: StructType) -> "pa.Schema":
-    """Convert a schema from Spark to Arrow"""
+def to_arrow_schema(
+    schema: StructType,
+    error_on_duplicated_field_names_in_struct: bool = False,
+    timestamp_utc: bool = True,
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/types.py:182]: Seems like we don't use this? `timestamp_utc`

File: python/pyspark/sql/pandas/types.py (hunk 21 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/types.py:143]: For Classic, it worked fine before. Duplicated field names in a struct already raised the correct PySpark error.

But for Connect, this condition was unhandled by Spark before, and was only caught by PyArrow.

For example, see the last few lines of the test `test_toArrow_duplicate_field_names` which tests this condition. Without enforcing no duplicate field names in structs, this is the error raised in that test:

```
pyarrow.lib.ArrowTypeError: struct fields don't match or are in the wrong order: Input fields: struct<x_0: string, x_1: int32> output fields: struct<x: string, x: int32>
```

With the code added here to enforce it, we see the correct error in Connect, the same as in Classic:

```
pyspark.errors.exceptions.base.UnsupportedOperationException: [DUPLICATED_FIELD_NAME_IN_ARROW_STRUCT] Duplicated field names in Arrow Struct are not allowed, got ['x', 'x']
```

File: python/pyspark/sql/pandas/types.py (hunk 23 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/types.py:182]: It is used in one place in the tests. See `test_createDataFrame_arrow_nested_timestamp`.

File: python/pyspark/sql/pandas/conversion.py (hunk 22 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/conversion.py:381]: I think this is the usual method that Python developers recommend for checking the type of an object when you don't know if the module is loaded. It looks like this method is used in many other places in the PySpark code.

One improvement would be to explicitly check the module also, like this:

```py
if (type(t).__module__, type(t).__name__) == ("pyarrow.lib", "Table"):
```

I am happy to use a different approach but I am not sure which approach would be an improvement. I think the approach you suggest will not work because `isinstance` expects the name the module was loaded as (usually `pa`) but the method to check if PyArrow is loaded (`pyarrow in sys.modules`) needs the library name.

Comment (ianmcook) [reply python/pyspark/sql/pandas/conversion.py:381]: Alternatively, I could add a separate definition for `createDataFrame` only for `data: "pa.Table"`. Let me know if you think that is preferable.

Comment (BryanCutler) [reply python/pyspark/sql/pandas/conversion.py:381]: I think it would be good to check the module also, although I'm not sure if there are any other variations to "pyarrow.lib"?

File: python/pyspark/sql/pandas/conversion.py [hunk 24]
```diff
@@ -360,12 +376,25 @@ def createDataFrame(  # type: ignore[misc]
...
+            # `data` is a PyArrow Table
+            from pyspark.sql.pandas.utils import require_minimum_pyarrow_version
+
+            require_minimum_pyarrow_version()
```
Comment (BryanCutler) [inline python/pyspark/sql/pandas/conversion.py:385]: Would it be good to check pa.Table with isinstance at this point?

**Review by BryanCutler — APPROVED** [review]
Looks great, thanks for doing this!

File: python/pyspark/sql/pandas/conversion.py (hunk 24 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/conversion.py:385]: Sure, I could add an assertion here just to be safe. I think the type hints in the function signature should prevent us from getting here if it's not a PyArrow Table but I'll add an assertion for good measure.

File: python/pyspark/sql/pandas/conversion.py (hunk 22 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/conversion.py:381]: Hmm yeah I'm not sure. Probably it's safer to add an `isinstance` assertion below as you suggest.

File: python/pyspark/sql/pandas/conversion.py (hunk 24 above)
Comment (ianmcook) [reply python/pyspark/sql/pandas/conversion.py:385]: Added in ee489f0.

**Comment by ianmcook** [comment]
Just rebased to master

**Comment by HyukjinKwon** [comment]
will merge once ci passes

**Comment by ianmcook** [comment]
Thanks @HyukjinKwon. After you merge I will rebase the small follow-up PR #46688 and prep it for review.

**Comment by HyukjinKwon** [comment]
Merged to master.
